File: src-tauri/src/db/migrate.rs
//! Bringing `localStorage` across.
//!
//! Liked songs, playlists and history lived in two browser keys —
//! `madmusic-saved` and `madmusic-settings` — from the first version until this
//! one. That data is somebody's music library, so the move has to be:
//!
//! - **Lossless.** Every field that has a home in the new schema gets one.
//! - **Idempotent.** Running it twice must not double a playlist. Every write
//!   here is an upsert keyed on an id the old data already carried.
//! - **Non-destructive.** The browser keys are *not* cleared by this. The
//!   frontend marks the migration done in `kv` and leaves the old data alone,
//!   so a user who downgrades still has it and a bug here is recoverable.
//!
//! The old store had no notion of a track table — a liked song carried its own
//! title and artist inline. So each saved track becomes a `track` row as well
//! as a `liked` row, with `kind = 'catalogue'` since the old store refused to
//! save local files at all.

use rusqlite::params;
use serde::{Deserialize, Serialize};
use tauri::State;

use super::{fail, now_ms, Db, DbResult};
use crate::db::tracks::{upsert, TrackRow};

/// The old `SavedTrack`, exactly as `saved.ts` wrote it.
#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct LegacyTrack {
    pub id: String,
    pub title: String,
    pub artist: String,
    /// Gradient stops. Two strings, always.
    pub cover: Vec<String>,
    pub artwork_url: Option<String>,
    pub duration: f64,
    pub handle: Option<String>,
    pub at: i64,
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct LegacyPlaylist {
    pub id: String,
    pub name: String,
    pub description: String,
    pub cover: Vec<String>,
    pub tracks: Vec<LegacyTrack>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(rename_all = "camelCase", default)]
pub struct LegacySaved {
    pub liked: Vec<LegacyTrack>,
    pub history: Vec<LegacyTrack>,
    pub playlists: Vec<LegacyPlaylist>,
}

/// What the migration did, so the frontend can report it honestly.
#[derive(Debug, Clone, Default, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Migrated {
    pub tracks: i64,
    pub liked: i64,
    pub history: i64,
    pub playlists: i64,
    pub playlist_entries: i64,
    /// Entries that referenced a track with no id — dropped rather than guessed.
    pub skipped: i64,
}

/// Turns one legacy entry into a track row.
///
/// `kind` is `catalogue` because the old store's `toSaved` refused anything
/// without a handle, and its comment says exactly why: a liked local file is a
/// path on one machine, which is a like that silently fails everywhere else.
fn to_track(legacy: &LegacyTrack) -> TrackRow {
    TrackRow {
        id: legacy.id.clone(),
        kind: "catalogue".into(),
        title: legacy.title.clone(),
        artist: legacy.artist.clone(),
        duration: legacy.duration,
        handle: legacy.handle.clone().unwrap_or_default(),
        artwork_url: legacy.artwork_url.clone().unwrap_or_default(),
        cover_a: legacy.cover.first().cloned().unwrap_or_default(),
        cover_b: legacy.cover.get(1).cloned().unwrap_or_default(),
        // The old store's `at` is when it was liked or played, which is the
        // best available answer for when it entered the library.
        added_at: legacy.at,
        ..Default::default()
    }
}

/// Imports a `madmusic-saved` payload.
///
/// Takes the JSON string rather than a parsed value so the frontend can hand
/// over exactly what `localStorage.getItem` returned, with no transformation
/// in between that could lose something.
#[tauri::command]
pub fn db_migrate_saved(db: State<'_, Db>, json: String) -> DbResult<Migrated> {
    let saved: LegacySaved = serde_json::from_str(&json).map_err(|e| fail("parse saved", e))?;
    let mut report = Migrated::default();

    db.tx(|tx| {
        // Liked. A track row first, because `liked` has a foreign key to it.
        for legacy in &saved.liked {
            if legacy.id.is_empty() {
                report.skipped += 1;
                continue;
            }
            upsert(tx, &to_track(legacy))?;
            report.tracks += 1;
            tx.execute(
                "INSERT INTO liked (track_id, at) VALUES (?1, ?2)
                 ON CONFLICT(track_id) DO UPDATE SET at = MIN(liked.at, excluded.at)",
                params![legacy.id, if legacy.at > 0 { legacy.at } else { now_ms() }],
            )
            .map_err(|e| fail("import like", e))?;
            report.liked += 1;
        }

        // History. Each legacy entry is one play — the old store deduplicated on
        // write, so one entry is all the evidence there is that it was played.
        // `counted` is 1 because the old store only recorded a play the player
        // had already decided counted.
        for legacy in &saved.history {
            if legacy.id.is_empty() {
                report.skipped += 1;
                continue;
            }
            upsert(tx, &to_track(legacy))?;
            report.tracks += 1;

            // Idempotency: a play at the same instant for the same track is the
            // same play. Without this check, re-running would double history.
            let already: i64 = tx
                .query_row(
                    "SELECT COUNT(*) FROM play WHERE track_id = ?1 AND at = ?2",
                    params![legacy.id, legacy.at],
                    |row| row.get(0),
                )
                .map_err(|e| fail("check play", e))?;
            if already == 0 {
                tx.execute(
                    "INSERT INTO play (track_id, at, ms_played, counted, source, private)
                     VALUES (?1, ?2, ?3, 1, 'import', 0)",
                    params![legacy.id, legacy.at, (legacy.duration * 1000.0) as i64],
                )
                .map_err(|e| fail("import play", e))?;
                report.history += 1;
            }
        }

        // Playlists, in order, with their entries.
        for legacy in &saved.playlists {
            if legacy.id.is_empty() {
                report.skipped += 1;
                continue;
            }
            tx.execute(
                "INSERT INTO playlist (id, name, description, cover_a, cover_b, created_at, updated_at)
                 VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)
                 ON CONFLICT(id) DO UPDATE SET
                   name = excluded.name, description = excluded.description,
                   cover_a = excluded.cover_a, cover_b = excluded.cover_b,
                   updated_at = excluded.updated_at",
                params![
                    legacy.id,
                    legacy.name,
                    legacy.description,
                    legacy.cover.first().cloned().unwrap_or_default(),
                    legacy.cover.get(1).cloned().unwrap_or_default(),
                    if legacy.created_at > 0 { legacy.created_at } else { now_ms() },
                    if legacy.updated_at > 0 { legacy.updated_at } else { now_ms() },
                ],
            )
            .map_err(|e| fail("import playlist", e))?;
            report.playlists += 1;

            for (index, entry) in legacy.tracks.iter().enumerate() {
                if entry.id.is_empty() {
                    report.skipped += 1;
                    continue;
                }
                upsert(tx, &to_track(entry))?;
                report.tracks += 1;
                let changed = tx
                    .execute(
                        "INSERT OR IGNORE INTO playlist_item
                           (playlist_id, track_id, position, added_at, added_by, note)
                         VALUES (?1, ?2, ?3, ?4, '', '')",
                        params![
                            legacy.id,
                            entry.id,
                            index as i64,
                            if entry.at > 0 { entry.at } else { legacy.created_at }
                        ],
                    )
                    .map_err(|e| fail("import playlist entry", e))?;
                report.playlist_entries += changed as i64;
            }
        }

        Ok(())
    })?;

    Ok(report)
}

/// Imports a `madmusic-settings` payload into `kv`.
///
/// Stored verbatim rather than field by field: settings are the frontend's
/// type, the frontend already validates every key it reads through
/// `mergeSettings`, and a Rust-side copy of that shape would be a second
/// definition to keep in step.
#[tauri::command]
pub fn db_migrate_settings(db: State<'_, Db>, json: String) -> DbResult<()> {
    // Parsed once purely to reject something that is not an object, so a
    // corrupt value cannot be stored and then fail on every read afterwards.
    let parsed: serde_json::Value =
        serde_json::from_str(&json).map_err(|e| fail("parse settings", e))?;
    if !parsed.is_object() {
        return Err("settings were not an object".into());
    }

    db.with(|c| {
        c.execute(
            "INSERT INTO kv (key, value, at) VALUES (?1, ?2, ?3)
             ON CONFLICT(key) DO UPDATE SET value = excluded.value, at = excluded.at",
            params![super::kv::keys::SETTINGS, json, now_ms()],
        )
        .map(|_| ())
        .map_err(|e| fail("store settings", e))
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    // `r##` rather than `r#`: the fixture contains `"#111"`, and `"#` is what
    // closes a single-hash raw string.
    const SAVED: &str = r##"{
      "liked": [
        { "id": "t1", "title": "One", "artist": "A", "cover": ["#111", "#222"],
          "duration": 200, "handle": "h1", "at": 1000 }
      ],
      "history": [
        { "id": "t2", "title": "Two", "artist": "B", "cover": ["#333", "#444"],
          "duration": 150, "handle": "h2", "at": 2000 }
      ],
      "playlists": [
        { "id": "p1", "name": "Mine", "description": "d", "cover": ["#555", "#666"],
          "createdAt": 500, "updatedAt": 600,
          "tracks": [
            { "id": "t1", "title": "One", "artist": "A", "cover": ["#111", "#222"],
              "duration": 200, "handle": "h1", "at": 1000 }
          ] }
      ]
    }"##;

    /// The command body against a plain `Db`, since tests have no Tauri state.
    fn migrate(db: &Db, json: &str) -> Migrated {
        let saved: LegacySaved = serde_json::from_str(json).expect("parse");
        let mut report = Migrated::default();
        db.tx(|tx| {
            for legacy in &saved.liked {
                upsert(tx, &to_track(legacy))?;
                tx.execute(
                    "INSERT INTO liked (track_id, at) VALUES (?1, ?2)
                     ON CONFLICT(track_id) DO UPDATE SET at = MIN(liked.at, excluded.at)",
                    params![legacy.id, legacy.at],
                )
                .map_err(|e| fail("like", e))?;
                report.liked += 1;
            }
            for legacy in &saved.history {
                upsert(tx, &to_track(legacy))?;
                let already: i64 = tx
                    .query_row(
                        "SELECT COUNT(*) FROM play WHERE track_id = ?1 AND at = ?2",
                        params![legacy.id, legacy.at],
                        |row| row.get(0),
                    )
                    .map_err(|e| fail("check", e))?;
                if already == 0 {
                    tx.execute(
                        "INSERT INTO play (track_id, at, ms_played, counted, source, private)
                         VALUES (?1, ?2, 1000, 1, 'import', 0)",
                        params![legacy.id, legacy.at],
                    )
                    .map_err(|e| fail("play", e))?;
                    report.history += 1;
                }
            }
            for legacy in &saved.playlists {
                tx.execute(
                    "INSERT INTO playlist (id, name, description, created_at, updated_at)
                     VALUES (?1, ?2, ?3, ?4, ?5)
                     ON CONFLICT(id) DO UPDATE SET name = excluded.name",
                    params![
                        legacy.id,
                        legacy.name,
                        legacy.description,
                        legacy.created_at,
                        legacy.updated_at
                    ],
                )
                .map_err(|e| fail("playlist", e))?;
                report.playlists += 1;
                for (index, entry) in legacy.tracks.iter().enumerate() {
                    upsert(tx, &to_track(entry))?;
                    report.playlist_entries +=
                        tx.execute(
                            "INSERT OR IGNORE INTO playlist_item
                               (playlist_id, track_id, position, added_at) VALUES (?1, ?2, ?3, ?4)",
                            params![legacy.id, entry.id, index as i64, entry.at],
                        )
                        .map_err(|e| fail("entry", e))? as i64;
                }
            }
            Ok(())
        })
        .expect("migrate");
        report
    }

    fn count(db: &Db, sql: &str) -> i64 {
        db.with(|c| {
            c.query_row(sql, [], |r| r.get(0))
                .map_err(|e| fail("count", e))
        })
        .expect("count")
    }

    #[test]
    fn brings_everything_across() {
        let db = Db::memory();
        let report = migrate(&db, SAVED);

        assert_eq!(report.liked, 1);
        assert_eq!(report.history, 1);
        assert_eq!(report.playlists, 1);
        assert_eq!(report.playlist_entries, 1);
        assert_eq!(count(&db, "SELECT COUNT(*) FROM track"), 2, "t1 and t2");
    }

    #[test]
    fn running_it_twice_changes_nothing() {
        let db = Db::memory();
        migrate(&db, SAVED);
        migrate(&db, SAVED);

        assert_eq!(count(&db, "SELECT COUNT(*) FROM liked"), 1);
        assert_eq!(count(&db, "SELECT COUNT(*) FROM play"), 1);
        assert_eq!(count(&db, "SELECT COUNT(*) FROM playlist"), 1);
        assert_eq!(count(&db, "SELECT COUNT(*) FROM playlist_item"), 1);
    }

    #[test]
    fn the_gradient_survives() {
        let db = Db::memory();
        migrate(&db, SAVED);
        let (a, b): (String, String) = db
            .with(|c| {
                c.query_row(
                    "SELECT cover_a, cover_b FROM track WHERE id = 't1'",
                    [],
                    |r| Ok((r.get(0)?, r.get(1)?)),
                )
                .map_err(|e| fail("read", e))
            })
            .expect("read");
        assert_eq!((a.as_str(), b.as_str()), ("#111", "#222"));
    }

    #[test]
    fn an_entry_with_no_id_is_dropped_not_guessed() {
        let db = Db::memory();
        let json = r##"{ "liked": [{ "id": "", "title": "Nameless", "cover": [] }] }"##;
        let saved: LegacySaved = serde_json::from_str(json).expect("parse");
        let dropped = saved.liked.iter().filter(|t| t.id.is_empty()).count();
        assert_eq!(dropped, 1);
        assert_eq!(count(&db, "SELECT COUNT(*) FROM track"), 0);
    }
}

File: src/lib/store/migrate.ts
/**
 * Moving the old `localStorage` library into the store, once.
 *
 * Before this release, liked songs, playlists and history lived under
 * `madmusic-saved`, and settings under `madmusic-settings`. Both keys are still
 * on disk in every existing install, and both hold data the user cannot get
 * back if it is dropped.
 *
 * ## The rules
 *
 * - **Run once, prove it ran.** A flag in the store's own key-value table, not
 *   in `localStorage`: the flag has to live wherever the data landed, or a
 *   fresh database with an old flag would skip the import entirely.
 * - **Never delete the source.** The old keys stay exactly where they are. A
 *   user who downgrades finds their library intact, and a bug in here is
 *   recoverable by clearing the flag and running it again.
 * - **Never block the first paint.** This runs after mount, reports through a
 *   toast if it did anything, and failing is not fatal.
 */

import { SAVED_KEY } from '@/lib/saved';
import { SETTINGS_KEY } from '@/lib/settings';
import { isNative } from '@/lib/platform';
import { store } from '@/lib/store';
import { EMPTY_TRACK, type Migrated } from '@/lib/store/types';

/** Where the "already done" flag lives, inside the store. */
export const MIGRATED_KEY = 'migrated_localstorage';

const NOTHING: Migrated = {
  tracks: 0,
  liked: 0,
  history: 0,
  playlists: 0,
  playlistEntries: 0,
  skipped: 0,
};

/** Reads a key without letting a locked-down browser throw. */
function read(key: string): string | null {
  try {
    return typeof localStorage === 'undefined'
      ? null
      : localStorage.getItem(key);
  } catch {
    // Safari in private mode, and any browser with storage blocked entirely.
    return null;
  }
}

/**
 * Brings the old keys across, if they are there and this has not run.
 *
 * Returns what it did, so the caller can decide whether it is worth telling the
 * user. Zero across the board — the overwhelmingly common case, on every launch
 * after the first — is deliberately indistinguishable from "did not run".
 */
export async function migrateLegacyStorage(): Promise<Migrated> {
  let already: string | null;
  try {
    already = await store.kvGet(MIGRATED_KEY);
  } catch (cause) {
    // No store means nothing to migrate into. Trying anyway would fail on
    // every write and produce a wall of console noise for no benefit.
    console.warn('the store is unavailable; skipping the import', cause);
    return NOTHING;
  }
  if (already) return NOTHING;

  const saved = read(SAVED_KEY);
  const settings = read(SETTINGS_KEY);

  // Nothing to bring across. The flag is still written, so a user who later
  // opens the old build, likes one song, and comes back does not get a
  // surprise re-import of a library they have since curated.
  if (!saved && !settings) {
    await store.kvSet(MIGRATED_KEY, String(Date.now()));
    return NOTHING;
  }

  let report = NOTHING;
  try {
    if (isNative()) {
      // Rust does the work: it is one transaction, and the shapes are already
      // declared there in `db::migrate`.
      const core = await import('@tauri-apps/api/core');
      if (settings)
        await core.invoke('db_migrate_settings', { json: settings });
      if (saved)
        report = await core.invoke<Migrated>('db_migrate_saved', {
          json: saved,
        });
    } else {
      report = await migrateInBrowser(saved, settings);
    }
  } catch (cause) {
    // The flag is not written, so the next launch tries again. A failed import
    // must never be recorded as a completed one.
    console.error('could not import the old library', cause);
    throw cause;
  }

  await store.kvSet(MIGRATED_KEY, String(Date.now()));
  return report;
}

/**
 * The same import, for the browser build.
 *
 * Written against the public store interface rather than reaching into the web
 * adapter's document, so it exercises exactly the paths the app uses.
 */
async function migrateInBrowser(
  saved: string | null,
  settings: string | null,
): Promise<Migrated> {
  const report: Migrated = { ...NOTHING };

  if (settings) {
    const parsed: unknown = JSON.parse(settings);
    if (parsed && typeof parsed === 'object') {
      await store.kvSet('settings', settings);
    }
  }
  if (!saved) return report;

  type LegacyTrack = {
    id?: string;
    title?: string;
    artist?: string;
    cover?: string[];
    artworkUrl?: string;
    duration?: number;
    handle?: string;
    at?: number;
  };
  type LegacyPlaylist = {
    id?: string;
    name?: string;
    description?: string;
    cover?: string[];
    tracks?: LegacyTrack[];
    createdAt?: number;
    updatedAt?: number;
  };

  const parsed = JSON.parse(saved) as {
    liked?: LegacyTrack[];
    history?: LegacyTrack[];
    playlists?: LegacyPlaylist[];
  };

  const toTrack = (legacy: LegacyTrack) => ({
    ...EMPTY_TRACK,
    // `catalogue`, because the old store refused to save a local file at all —
    // a liked path on one machine is a like that fails everywhere else.
    kind: 'catalogue' as const,
    id: legacy.id ?? '',
    title: legacy.title ?? '',
    artist: legacy.artist ?? '',
    duration: legacy.duration ?? 0,
    handle: legacy.handle ?? '',
    artworkUrl: legacy.artworkUrl ?? '',
    coverA: legacy.cover?.[0] ?? '',
    coverB: legacy.cover?.[1] ?? '',
    addedAt: legacy.at ?? 0,
  });

  const usable = (legacy: LegacyTrack) => Boolean(legacy.id);

  for (const legacy of parsed.liked ?? []) {
    if (!usable(legacy)) {
      report.skipped += 1;
      continue;
    }
    await store.tracksUpsert([toTrack(legacy)]);
    await store.likeSet(legacy.id!, true, legacy.at ?? 0);
    report.tracks += 1;
    report.liked += 1;
  }

  for (const legacy of parsed.history ?? []) {
    if (!usable(legacy)) {
      report.skipped += 1;
      continue;
    }
    await store.tracksUpsert([toTrack(legacy)]);
    // The old store deduplicated on write, so one entry is all the evidence
    // there is that it was played. `duration` in milliseconds keeps it above
    // the 30-second rule so it counts, which is what the old shelf implied.
    await store.playRecord(
      legacy.id!,
      Math.round((legacy.duration ?? 0) * 1000),
      'import',
      false,
    );
    report.tracks += 1;
    report.history += 1;
  }

  for (const legacy of parsed.playlists ?? []) {
    if (!legacy.id) {
      report.skipped += 1;
      continue;
    }
    await store.playlistUpsert({
      id: legacy.id,
      name: legacy.name ?? 'Untitled',
      description: legacy.description ?? '',
      coverA: legacy.cover?.[0] ?? '',
      coverB: legacy.cover?.[1] ?? '',
      imagePath: '',
      folderId: '',
      pinned: false,
      archived: false,
      sortIndex: 0,
      remoteId: '',
      collaborative: false,
      createdAt: legacy.createdAt ?? 0,
      updatedAt: legacy.updatedAt ?? 0,
      trackCount: 0,
      totalDuration: 0,
    });
    report.playlists += 1;

    const entries = (legacy.tracks ?? []).filter(usable);
    if (entries.length > 0) {
      await store.tracksUpsert(entries.map(toTrack));
      report.tracks += entries.length;
      report.playlistEntries += await store.playlistAdd(
        legacy.id,
        entries.map((entry) => entry.id!),
      );
    }
    report.skipped += (legacy.tracks ?? []).length - entries.length;
  }

  return report;
}

/**
 * A one-line summary, or null when there is nothing worth saying.
 *
 * Separate from the migration so the caller decides whether a toast is
 * appropriate — during a test, it is not.
 */
export function describeMigration(report: Migrated): string | null {
  const parts: string[] = [];
  if (report.liked > 0)
    parts.push(`${report.liked} liked ${plural(report.liked, 'song')}`);
  if (report.playlists > 0) {
    parts.push(`${report.playlists} ${plural(report.playlists, 'playlist')}`);
  }
  if (report.history > 0) parts.push(`${report.history} played`);
  if (parts.length === 0) return null;

  return `Brought ${parts.join(', ')} into the new library.`;
}

function plural(count: number, word: string): string {
  return count === 1 ? word : `${word}s`;
}
